Nodejs/Product/Nodejs/SharedProject/FileChangeManager.cs

```diff
@@ -1,92 +1,145 @@
 using Microsoft.VisualStudio;
 using Microsoft.VisualStudio.Shell.Interop;
 using IServiceProvider = System.IServiceProvider;
 
 namespace Microsoft.VisualStudioTools.Project
 {
+    public sealed class FolderChangedEventArgs : EventArgs
+    {
+        public readonly string FolderName;
+        public readonly string FileName;
+        public _VSFILECHANGEFLAGS FileChangeFlag;
+
+        public FolderChangedEventArgs(string folderName, string fileName, _VSFILECHANGEFLAGS fileChangeFlag)
+        {
+            this.FolderName = folderName;
+            this.FileName = fileName;
+            this.FileChangeFlag = fileChangeFlag;
+        }
+    }
+
     /// <summary>
-    /// This object is in charge of reloading nodes that have file monikers that can be listened to changes
+    /// This object is in charge of watching for changes to files and folders.
     /// </summary>
-    internal class FileChangeManager : IVsFileChangeEvents
+    internal sealed class FileChangeManager
     {
-        /// <summary>
-        /// Defines a data structure that can link a item moniker to the item and its file change cookie.
-        /// </summary>
-        private struct ObservedItemInfo
+        private sealed class FileChangeEvents : IVsFreeThreadedFileChangeEvents
         {
-            /// <summary>
-            /// Defines the id of the item that is to be reloaded.
-            /// </summary>
-            private uint itemID;
+            private readonly FileChangeManager fileChangeManager;
 
-            /// <summary>
-            /// Defines the file change cookie that is returned when listening on file changes on the nested project item.
-            /// </summary>
-            private uint fileChangeCookie;
+            public FileChangeEvents(FileChangeManager fileChangeManager)
+            {
+                this.fileChangeManager = fileChangeManager;
+            }
 
             /// <summary>
-            /// Defines the nested project item that is to be reloaded.
+            /// Called when one of the file have changed on disk.
             /// </summary>
-            internal uint ItemID
+            /// <param name="numberOfFilesChanged">Number of files changed.</param>
+            /// <param name="filesChanged">Array of file names.</param>
+            /// <param name="flags">Array of flags indicating the type of changes. See _VSFILECHANGEFLAGS.</param>
+            /// <returns>If the method succeeds, it returns S_OK. If it fails, it returns an error code.</returns>
+            public int FilesChanged(uint numberOfFilesChanged, string[] filesChanged, uint[] flags)
             {
-                get
+                if (filesChanged == null)
+                {
+                    throw new ArgumentNullException(nameof(filesChanged));
+                }
+
+                if (flags == null)
                 {
-                    return this.itemID;
+                    throw new ArgumentNullException(nameof(flags));
                 }
```

Comment:
These checks seem only to detect programmer errors in VS.  Would it make more sense to assert?  If there's actually a danger of bad input, we should also check that `numberOfFilesChanged` isn't out of bounds.

Reply:
kept the pattern used in the `FilesChanged` method